### ISSUE TITLE
1080p

### ISSUE THREAD

**Issue by primaeval** [vonH/plugin.video.iplayerwww#47]
Have you had a look at the Plex Media Server iplayer addon? They have found the 1080p streams, not that I can play them reliably ;) They also use some interesting xpath searches instead of regex or soup.

**Comment by CaptainTK** [comment]
Very interesting hint. They use an entirely different stream source, but the actual content seems to be the same as in this add-on. I also noted that their source claims to be 1080p, but I could only find 720p streams in the m3u8 file. Maybe I was just testing a programme which was not suitable.

What's more, the same kind of 1080p also appeared in the stream sources here, it is just being ignored by ParseStreams. Should be an easy extension, so no need to switch the source to the same they use.

Did you find any programmes which actually played in 1080p?

XPath sounds interesting, but may not work the way this add-on is set up. In the Plex add-on, they use different sources for the programme information which provide XML or JSON responses. These sources look more reverse engineered than just scraping websites, which sounds kind of fishy to me. No idea, how they found this, but I wouldn't want to use it myself.

**Comment by primaeval** [comment]
I used the Plexbmc addon to play the streams in 1080p in Kodi. I was careful not to switch on transcoding so I think they were proper 1080p. Maybe it was live tv.

There are quite a few different smart tv boxes around now, maybe that was their source. get_iplayer grabs xml schedules which are nicely formatted but only go back 2 weeks.

**Comment by primaeval** [comment]
I just did a full check. As far as I can see in Kodi and Plex logs it could play back John Bishop's show in 1080p without transcoding but only for a few seconds before buffering on my connection. 
Their title naming is screwed up and live streams won't play for me. Maybe they are not so clever ;)

**Comment by CaptainTK** [comment]
Thanks for the John Bishop hint.

Good news is: I was able to find the 1080p stream for that show.

Bad news is: I found a lot of other streams as well. Lots of new resolutions, bitrates, audio formats, very confusing.

To sum up: 1080p is possible, I know how, but it will need more rewriting in the add-on to handle all of the different bitrate options.

**Comment by primaeval** [comment]
Can Kodi work out the best bitrate if you send it an m3u8 with different streams in? Some desktop players can do that. A lot of iplayerwww users can't handle the high default bitrate. That would be a good function to implement if you can work it out.

**Comment by CaptainTK** [comment]
Yes, that would be perfect, but AFAIK isn't supported, see http://forum.kodi.tv/showthread.php?tid=129912%29, for some more information

**Comment by primaeval** [comment]
Maybe it's best to have the stream selection as default behaviour so users can find the best bitrate before setting auto on.

**Comment by CaptainTK** [comment]
I agree, but it was a requirement to make this the non-default to get on the official repo, so I won't change it back now.

**Comment by primaeval** [comment]
That's tricky then. I've had to help a few people with the bitrate. Maybe a popup on first use, although I hate nag screens.

**Comment by willrogers** [comment]
I had a problem with the default bitrate being too high so using the plugin was very irritating until I found the settings page.

Couldn't you make the default more conservative?

**Comment by CaptainTK** [comment]
Thanks for your comment, although it is not related to this issue.

Making the default more conservative would just shift the problem to other users, who would then suffer from poor quality. My line can even handle 1080p at day, but at night my is is sometimes too slow for SD streams, so I need to select manually.

The only real solution would be for Kodi to support dynamic switching of bitrates according to buffer level. Unfortunately, this is a Kodi function, not an add-on function and it seems that it is not supported in Kodi, see link above.

Long story short: Whatever the default is: It is wrong for somebody. Perhaps it would help to add an FAQ section explaining these kind of things to the first forum post.

**Comment by CaptainTK** [comment]
Back to the 1080p idea: I have decided not to integrate 1080p for the time being, for two reasons:

1. If 1080p was included, it would become the new default. Since users already struggle with 720p as a default, this may lead to more issues. The 1080p streams have a bitrate close to 9 Mbit/s.
2. The 1080p streams don't seem to be official yet. I would not like to support them before the Beeb does.

This issue should stay open until 1080p can and will be integrated in the future.

**Comment by primaeval** [comment]
The 1080p streams are nowhere to be found. Might as well close this.

**Comment by t0mtee** [comment]
This issue is ancient but I want to ask if this is still a possibility?

**Comment by Liqianyu** [comment]
https://www.bbc.co.uk/iplayer/help/questions/supported-devices/1080-streams
https://www.bbc.co.uk/iplayer/help/questions/features/uhd-connected-tv

According to a comment from reddit, to get 4K you need to extract a client certificate from a 4K-enabled device (TV, set-top box, etc.).
It seems a lot easier to get 1080p content though, and there have been many public implementations and discussions.
https://github.com/get-iplayer/get_iplayer
https://github.com/ytdl-org/youtube-dl/issues/30136

I don't think supporting 1080p should require too complex a modification, I'll look into it.

**Comment by CaptainTK** [comment]
Yes, getting 1080p should be relatively easy. However, the type of modification required is exactly what made the add-on instable for a long time. There was fixes over fixes over fixes. Basically, the add-on was more broken then working. Also, it made debugging issues extremely difficult.

Since I modified the add-on to work exactly as the website does, it has become extremely stable. Only larger changes on the website created a couple of weeks of turmoil until everything settled again for months.

Based on this experience, I will not add 1080p to the add-on unless it is also offered on the website. The limitation to 720p is a small price to pay for stability.

Feel free to create your own branch and add whatever you like. However, as long as I maintain this add-on, I will not include code that emulates devices other than a standard web browser or uses other mediaselectors than the website.

**Comment by Liqianyu** [comment]
Simply put, to support 1080p, you need the add-on to modify the fetched MPD file by changing the id="video=5070000" in it, to id="video=12000000", and then play it.
Manually modifying MPD and creating STRM can play iPlayer FHD(1080p).

I suspect that iPlayer may never (at least not for a long time) put 1080p on a web browser.
But as @CaptainTK said, this modification is unstable. It's a hacking technique.

An analog TV client to get the correct MPD file would be a better option, but that would require more work now.

**Comment by t0mtee** [comment]
> Simply put, to support 1080p, you need the add-on to modify the fetched MPD file by changing the id="video=5070000" in it, to id="video=12000000", and then play it. Manually modifying MPD and creating STRM can play iPlayer FHD(1080p).
> 
> I suspect that iPlayer may never (at least not for a long time) put 1080p on a web browser. But as @CaptainTK said, this modification is unstable. It's a hacking technique.
> 
> An analog TV client to get the correct MPD file would be a better option, but that would require more work now.

Sorry for the very late response, do you know how I can make this change? I'd like to at least test it, and see how unstable it is on my setup. Thanks!
